lib/src/main/java/com/basistheory/android/view/mask/Mask.kt

```diff
@@ -0,0 +1,76 @@
+package com.basistheory.android.view.mask
+
+internal class Mask(
+    private val mask: List<Any>,
+) {
+    private val maskPattern =
+        convertMaskPatternToString(
+            mask,
+            "_"
+        )
+
+    private val placeholder = maskPattern.getMostRepeatedChar()
+
+    fun apply(text: CharSequence, action: Action): MaskResult {
+        val source = text.iterator()
+        val maskedValue = mutableListOf<Char>()
+        val unMaskedValue = mutableListOf<Char>()
+        var currentChar = source.nextOrNull()
+
+        loop@ for ((index, value) in maskPattern.withIndex()) {
+            if (value == currentChar && value != placeholder) {
+                maskedValue.add(value)
+                currentChar = source.nextOrNull()
+            } else if (value == placeholder) {
+                currentChar = source.nextMaskChar(currentChar)
+                val originalValue = mask[index]
+                if (currentChar == null) {
+                    break@loop // do not render placeholders
+                } else if (originalValue is Regex && !originalValue.matches(currentChar.toString())) {
+                    break@loop
+                } else {
+                    maskedValue.add(currentChar)
+                    unMaskedValue.add(currentChar)
+                    currentChar = source.nextOrNull()
+                }
+            } else {
+                if (currentChar == null && action == Action.DELETE) break@loop
+                maskedValue.add(value)
+            }
+        }
+
+        var maskedText = maskedValue.joinToString("")
+        var unMaskedText = unMaskedValue.joinToString("")
+        val isDirty = unMaskedText.isNotEmpty()
+        val isDone = maskedText.length == maskPattern.length && !maskedText.contains(placeholder)
+
+        if (maskPattern.startsWith(maskedText)) {
+            maskedText = ""
+            unMaskedText = ""
+        }
+
+        return MaskResult(
+            maskedText, unMaskedText, isDone, isDirty
+        )
+    }
+
+    private fun convertMaskPatternToString(
+        mask: List<Any>,
+        placeholderChar: String
+    ): String = mask.map { if (it is Regex) placeholderChar else it }.joinToString("")
```

Comment:
I think `.joinToString("")` may be equivalent to `.toString()` here.

This also has me wondering if we need to support mixed patterns with different placeholders - we don't here but it seems like we don't either in web today, e.g. use a mix of digits and letters, with digits masked as '#' and letters masked as 'X' (ABC-1234 would have the placeholder text XXX-####).